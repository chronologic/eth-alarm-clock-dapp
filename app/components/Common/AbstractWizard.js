import React from 'react';

import { NAVIGATION_STEPS, PROPERTIES as ALL_PROPERTIES } from '../lib/consts';
export default class AbstractWizard extends React.Component {
  constructor(activeStepKey, activeApp, props) {
    super(props);

    this.activeStepKey = activeStepKey;
    this.activeApp = activeApp;

    this.activeStep = NAVIGATION_STEPS[this.activeApp][this.activeStepKey];
    if (!this.activeStep) {
      throw new Error('No steps with key', this.activeStepKey);
    }

    if (!this.activeStep.propertyKeys) {
      throw new Error('No propertyKeys for step', this.activeStep);
    }

    this.properties = ALL_PROPERTIES.reduce(
      (result, value) =>
        this.activeStep.propertyKeys.includes(value.name)
          ? { ...result, [value.name]: value }
          : result,
      {}
    );
  }

  async componentDidMount() {
    const { web3Service } = this.props;
    await web3Service.awaitInitialized();
    await this.validatePrevState();
  }

  web3web3Disabled(web3Service) {
    return (
      !web3Service.connectedToMetaMask ||
      !(typeof web3Service.accounts !== 'undefined' && web3Service.accounts.length > 0)
    );
  }
}
