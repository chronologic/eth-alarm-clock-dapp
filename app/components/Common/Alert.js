import React, { Component } from 'react';
import PropTypes from 'prop-types';

const Titles = {
  danger: 'Error',
  info: 'Info',
  warning: 'Warning',
  success: 'Done'
};

class Alert extends Component {
  render() {
    const type = this.props.type || 'danger';
    const close = typeof this.props.close !== 'undefined' ? this.props.close : true;
    const callToAction = this.props.action || null;
    return (
      <div className={'d-flex flex-nowrap alert alert-' + type} role="alert">
        <div className="d-flex flex-wrap flex-stretch flex-sm-nowrap justify-content-center justify-content-sm-start align-items-center ">
          <div className="flex-stretch">
            <strong>{Titles[type]}: </strong>
            {this.props.msg}
          </div>
          {callToAction && (
            <div className="d-inline-block ml-0 ml-sm-2 mt-2 mt-sm-0">{callToAction}</div>
          )}
        </div>
        <div>{close && <button className="close" data-dismiss="alert" />}</div>
      </div>
    );
  }
}

Alert.propTypes = {
  msg: PropTypes.string,
  type: PropTypes.string,
  close: PropTypes.any,
  action: PropTypes.any
};

export default Alert;
