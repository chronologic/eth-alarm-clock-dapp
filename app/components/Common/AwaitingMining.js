import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { inject, observer } from 'mobx-react';
import { services } from '../../services';
import Coder from 'web3/lib/solidity/coder';
import { PacmanLoader } from 'react-spinners';

const loaderConfig = {
  color: '#21ffff'
};

const mineDestinations = {
  scheduler: {
    path: 'transactions',
    prop: 'newContract',
    logEventHex: services.eacService.Constants.NEWREQUESTLOG,
    logEventTypes: ['address'],
    nextParameterPosition: 0
  }
};

@inject('web3Service')
@observer
class AwaitingMining extends Component {
  constructor(props) {
    super(props);

    const { type } = this.props.match.params;

    this.state = {
      waitType: type,
      transactionHash: '',
      newContract: '',
      deploying: false,
      minning: false
    };
  }

  async loadUp() {
    const { web3Service } = this.props;
    const {
      web3Service: { web3 }
    } = this.props;
    const { hash, type } = this.props.match.params;
    let unmined = true;
    let unconfirmed = true;

    await web3Service.awaitInitialized();

    if (web3.isAddress(hash)) {
      this.setState({ newContract: hash });
    } else {
      this.setState({ transactionHash: hash });
    }

    if (this.state.newContract) {
      return this.next();
    }

    if (this.state.transactionHash) {
      const { transactionHash } = this.state;

      while (unmined) {
        let txReceipt = await web3Service.fetchReceipt(transactionHash);
        if (!txReceipt) {
          this.setState({ deploying: true });
          await web3Service.trackTransaction(transactionHash);
        } else {
          this.setState({ deploying: false });
          unmined = false;
        }
      }

      while (unconfirmed) {
        const confirmations = await web3Service.fetchConfirmations(transactionHash);
        if (confirmations < 4) {
          this.setState({ minning: true });
        } else {
          this.setState({ minning: false });
          unconfirmed = false;
        }
      }

      if (mineDestinations[type].logEventTypes && mineDestinations[type].logEventHex) {
        const log = await web3Service.fetchLog(transactionHash, mineDestinations[type].logEventHex);
        const data = log.data.substring(2); //truncate data for decoding
        const args = Coder.decodeParams(mineDestinations[type].logEventTypes, data);
        let newSate = {};
        newSate[mineDestinations[type].prop] = args[mineDestinations[type].nextParameterPosition];
        this.setState(newSate);
      }

      this.next();
    }
  }

  getDestination() {
    const { type } = this.props.match.params;
    return '/' + mineDestinations[type].path + '/' + this.state[mineDestinations[type].prop];
  }

  next() {
    const { history } = this.props;
    const destination = this.getDestination();
    history.push({
      pathname: destination
    });
  }

  async componentDidMount() {
    await this.loadUp();
  }

  render() {
    const {
      web3Service: { explorer }
    } = this.props;
    const { transactionHash, newContract } = this.state;

    return (
      <div id="awaitingMining" className="container padding-25 sm-padding-10 horizontal-center">
        {this.state.deploying && <h1 className="view-title">Deploying</h1>}
        {this.state.minning && <h1 className="view-title">Awaiting Mining</h1>}
        {!this.state.deploying && !this.state.minning && <h1 className="view-title"> ... </h1>}
        <div className="card card-body">
          <div className="tabs-content">
            <div className="loader">
              <PacmanLoader {...Object.assign({ loading: true }, loaderConfig)} />
            </div>
            {this.state.transactionHash && (
              <p className="horizontal-center">
                Transaction Hash: <br />
                <a target="_blank" href={explorer + '/tx/' + transactionHash}>
                  {' '}
                  {this.state.transactionHash}{' '}
                </a>
              </p>
            )}
            {this.state.newContract && (
              <p className="horizontal-center">
                Contract Address: <br />
                <a target="_blank" href={explorer + '/address/' + newContract}>
                  {' '}
                  {this.state.newContract}{' '}
                </a>
              </p>
            )}
          </div>
        </div>
      </div>
    );
  }
}

AwaitingMining.propTypes = {
  match: PropTypes.any,
  web3Service: PropTypes.any,
  transactionStore: PropTypes.any,
  history: PropTypes.object.isRequired
};

export default AwaitingMining;
