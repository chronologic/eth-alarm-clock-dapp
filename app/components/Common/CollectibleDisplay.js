import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { inject } from 'mobx-react';

@inject('tokenHelper')
class CollectibleDisplay extends Component {
  getCollectibleIdDisplay() {
    const { collectibleId } = this.props;

    if (!collectibleId) {
      return '';
    }

    let display = collectibleId;

    if (typeof collectibleId !== 'string' && collectibleId.toString) {
      display = collectibleId.toString();
    }

    return display;
  }

  render() {
    const { onlyText, tokenAddress, tokenHelper, tokenName } = this.props;

    const collectibleId = this.getCollectibleIdDisplay();

    const token = tokenHelper.getTokenConfig(tokenAddress);

    if (!onlyText && token && token.imagePath) {
      return (
        <img
          src={tokenHelper.getTokenImagePath(token, collectibleId)}
          style={{ height: token.imageHeight, width: token.imageWidth }}
        />
      );
    }

    if (tokenName) {
      return (
        <span>
          {tokenName} (NFT-ERC721) with id: {collectibleId}
        </span>
      );
    }

    return <span>NFT-ERC721 with id: {collectibleId}</span>;
  }
}

CollectibleDisplay.propTypes = {
  collectibleId: PropTypes.any,
  onlyText: PropTypes.any,
  tokenAddress: PropTypes.any,
  tokenHelper: PropTypes.any,
  tokenName: PropTypes.any
};

export default CollectibleDisplay;
