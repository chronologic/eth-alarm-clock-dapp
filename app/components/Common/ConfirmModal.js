import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ConfirmModal extends Component {
  render() {
    return (
      <div
        className="modal fade stick-up"
        id={`${this.props.modalName}Modal`}
        tabIndex="-1"
        role="dialog"
        aria-labelledby={`${this.props.modalName}ModalLabel`}
        aria-hidden="true"
      >
        <div className="modal-dialog">
          <div className="modal-content">
            <div className="modal-header clearfix text-left separator">
              <button type="button" className="close" data-dismiss="modal" aria-hidden="true">
                <i className="pg-close fs-14" />
              </button>
              <h3 className="timenode-modal-title m-0">{this.props.modalTitle}</h3>
            </div>
            <div className="modal-body text-dark">
              <hr />
              <p>{this.props.modalBody}</p>
            </div>
            <div className="modal-footer">
              <div className="row">
                <div className="col-md-6">
                  <button
                    className="btn btn-light btn-block"
                    type="button"
                    data-dismiss="modal"
                    onClick={this.props.onCancel}
                  >
                    Cancel
                  </button>
                </div>
                <div className="col-md-6">
                  <button
                    className="btn btn-primary btn-block"
                    type="button"
                    data-dismiss="modal"
                    onClick={this.props.onConfirm}
                  >
                    <strong>Confirm</strong>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

ConfirmModal.propTypes = {
  modalName: PropTypes.string,
  modalTitle: PropTypes.string,
  modalBody: PropTypes.string,
  onConfirm: PropTypes.func,
  onCancel: PropTypes.func
};

export default ConfirmModal;
