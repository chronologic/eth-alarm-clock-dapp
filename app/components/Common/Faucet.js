import React from 'react';
import PropTypes from 'prop-types';
import { inject, observer } from 'mobx-react';
import Bb from 'bluebird';
import { BeatLoader } from 'react-spinners';
import { showNotification } from '../../services/notification';
import MetamaskComponent from '../Common/MetamaskComponent';

const Eth = 1e18;

@inject('web3Service')
@observer
class Faucet extends MetamaskComponent {
  constructor(props) {
    super(props);
    this.useFaucet = this.useFaucet.bind(this);
  }

  state = {
    loaded: false,
    defaultAccount: '',
    faucetAddress: '',
    dayTokenAddress: '',
    faucetBalance: 0,
    waitTime: 3600,
    waitTimeLeft: 0,
    lastUsed: ''
  };

  instance = {};

  get networkHasFaucet() {
    return Boolean(this.state.faucetAddress && this.state.faucetAddress != '0x0');
  }

  get waitTimeLeft() {
    const now = new Date().getTime();
    const nextAvailableTIme = (this.state.lastUsed + this.state.waitTime) * 1000;
    return now > nextAvailableTIme ? 0 : nextAvailableTIme - now;
  }

  get isEligible() {
    return (
      this.isWeb3Usable &&
      this.state.loaded &&
      this.networkHasFaucet &&
      this.state.defaultAccount &&
      this.waitTimeLeft == 0
    );
  }

  get printWaitTime() {
    let waitMins = 0;
    let waitSecs = 0;
    const Min = 60 * 1000;
    const waitTimeLeft = this.state.waitTimeLeft;
    if (waitTimeLeft > 0) {
      waitMins = Math.floor(waitTimeLeft / Min);
      waitSecs = Math.ceil((waitTimeLeft % Min) / 1000);
      this.startWaitTimeTimeout();
    }
    return `${waitMins} min(s) : ${waitSecs} sec(s)`;
  }

  setWaitTime() {
    this.setState({ waitTimeLeft: this.waitTimeLeft });
  }

  startWaitTimeTimeout() {
    const TIMEOUT = 1000;
    this.timeout = setTimeout(() => this.setWaitTime(), TIMEOUT);
  }

  startInterval() {
    const INTERVAL = 30000;
    this.interval = setInterval(() => this.loadInfo(), INTERVAL);
  }

  cancelInterval() {
    clearInterval(this.interval);
    clearTimeout(this.timeout);
  }

  restartInterval() {
    this.cancelInterval();
    this.startInterval();
  }

  async loadInfo() {
    this.setState({ loaded: false });
    const { web3Service } = this.props;
    await web3Service.init();

    const { accounts } = web3Service;
    this.setState({
      defaultAccount: accounts[0],
      faucetAddress: web3Service.network.dayFaucetAddress,
      dayTokenAddress: web3Service.network.dayTokenAddress
    });

    if (!this.isWeb3Usable || !this.state.faucetAddress) {
      return;
    }

    const {
      web3Service: { web3 }
    } = this.props;

    const faucetAbi = web3Service.network.dayFaucetAbi;
    this.instance = web3.eth.contract(faucetAbi).at(this.state.faucetAddress);

    this.setState({
      faucetBalance: Number(
        await Bb.fromCallback(callback => this.instance.getTokensBalance(callback))
      ),
      lastUsed: Number(
        await Bb.fromCallback(callback =>
          this.instance.lastRequest(this.state.defaultAccount, callback)
        )
      ),
      waitTime: Number(await Bb.fromCallback(callback => this.instance.waitTime(callback))),
      allowedTokens: Number(
        await Bb.fromCallback(callback => this.instance.allowedTokens(callback))
      )
    });

    this.setWaitTime();
    this.setState({ loaded: true });
  }

  useFaucet = async event => {
    const { target } = event;
    const {
      web3Service: { explorer }
    } = this.props;

    target.disabled = true;
    let transaction;
    try {
      transaction = await Bb.fromCallback(callback =>
        this.instance.useFaucet({ from: this.state.defaultAccount }, callback)
      );
      showNotification(
        `Transaction successful \r\n <a target='_blank' href='${explorer +
          'tx/' +
          transaction}'> ${transaction}<a>`,
        'success'
      );
      await this.restartInterval();
    } catch (e) {
      showNotification(`The transaction was unsuccessful.`);
    }
  };

  hrefProps() {
    return {
      target: '_blank',
      rel: 'noopener noreferrer'
    };
  }

  async componentDidMount() {
    super.componentDidMount();
    await this.loadInfo();
    this.startInterval();
  }

  componentWillUnmount() {
    super.componentWillUnmount();
    this.cancelInterval();
  }

  render() {
    const { web3Service } = this.props;
    const explorer = web3Service.explorer;

    const dayTokenAddressProps = this.hrefProps();
    const faucetAddressProps = this.hrefProps();
    const yourAddressProps = this.hrefProps();

    if (explorer) {
      dayTokenAddressProps.href = explorer + 'address/' + this.state.dayTokenAddress;
      faucetAddressProps.href = explorer + 'address/' + this.state.faucetAddress;
      yourAddressProps.href = explorer + 'address/' + this.state.defaultAccount;
    }

    return (
      <div className="container subsection">
        <h1 className="view-title">DAY Token Faucet</h1>
        <div className="card card-body">
          <div id="faucet" className="tab-content">
            <div className="tab-pane active show">
              <p>Get test DAY tokens on a testnet of your choice.</p>

              <div className="row">
                <div className="col-md-8 offset-md-2">
                  <div className="row my-3">
                    <div className="col-md-6 text-sm-right">
                      <strong>Testnet Network</strong>
                    </div>
                    <div className="col-md-6 text-left">
                      {this.state.loaded || web3Service.network ? (
                        web3Service.network.name
                      ) : (
                        <BeatLoader />
                      )}
                    </div>
                  </div>

                  <div className="row my-3">
                    <div className="col-md-6 text-sm-right">
                      <strong>Token Address</strong>
                    </div>
                    <div className="col-md-6 text-left">
                      {this.state.loaded || this.state.dayTokenAddress ? (
                        <a {...dayTokenAddressProps}>{this.state.dayTokenAddress}</a>
                      ) : (
                        <BeatLoader />
                      )}
                    </div>
                  </div>

                  <div className="row my-3">
                    <div className="col-md-6 text-sm-right">
                      <strong>Faucet Address</strong>
                    </div>
                    <div className="col-md-6 text-left">
                      {this.state.loaded || this.state.faucetAddress ? (
                        <a {...faucetAddressProps}>{this.state.faucetAddress}</a>
                      ) : (
                        <BeatLoader />
                      )}
                    </div>
                  </div>

                  <div className="row my-3">
                    <div className="col-md-6 text-sm-right">
                      <strong>Faucet Balance</strong>
                    </div>
                    <div className="col-md-6 text-left">
                      {this.state.loaded ||
                      (this.state.faucetAddress &&
                        typeof this.state.faucetBalance !== 'undefined') ? (
                        this.state.faucetBalance > 0 ? (
                          this.state.faucetBalance / Eth
                        ) : (
                          0
                        )
                      ) : (
                        <BeatLoader />
                      )}
                    </div>
                  </div>

                  <div className="row my-3">
                    <div className="col-md-6 text-sm-right">
                      <strong>Your Wallet Address</strong>
                    </div>
                    <div className="col-md-6 text-left">
                      {this.state.loaded || this.state.defaultAccount ? (
                        <a {...yourAddressProps}>{this.state.defaultAccount}</a>
                      ) : (
                        <BeatLoader />
                      )}
                    </div>
                  </div>

                  <div className="row my-3">
                    <div className="col-md-6 text-sm-right">
                      <strong>Remaining Wait Time</strong>
                    </div>
                    <div className="col-md-6 text-left">
                      {this.state.loaded ||
                      (this.state.defaultAccount && typeof this.state.lastUsed !== 'undefined') ? (
                        this.printWaitTime
                      ) : (
                        <BeatLoader />
                      )}
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <button
              className="btn btn-primary pull-right btn-lg"
              disabled={!this.isEligible || this.state.faucetBalance < this.state.allowedTokens}
              onClick={this.useFaucet}
            >
              Get Testnet DAY
            </button>
          </div>
        </div>
      </div>
    );
  }
}

Faucet.propTypes = {
  web3Service: PropTypes.any
};

export default Faucet;
