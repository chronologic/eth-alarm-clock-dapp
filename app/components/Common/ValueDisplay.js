import { Component } from 'react';
import PropTypes from 'prop-types';
import { inject } from 'mobx-react';

const ETHER_UNITS_VALUES_MAPPING = {
  WEI: 1,
  MWEI: 1000000,
  ETH: 1000000000000000000
};

@inject('web3Service')
export class ValueDisplay extends Component {
  render() {
    const { priceInWei } = this.props;

    if (!priceInWei) {
      return null;
    }

    return this._humanizeCurrencyDisplay(priceInWei);
  }

  /**
   * @private
   */
  _humanizeCurrencyDisplay(priceInWei) {
    let unit = 'ETH';

    if (!priceInWei) {
      return null;
    }

    const priceAsNumber = priceInWei.toNumber();

    let display = priceAsNumber;

    if (priceAsNumber < ETHER_UNITS_VALUES_MAPPING.MWEI && priceAsNumber > 0) {
      unit = 'WEI';
    } else {
      display = priceInWei.div(ETHER_UNITS_VALUES_MAPPING.ETH).toFixed();
      unit = 'ETH';
    }

    return `${display} ${unit}`;
  }
}

ValueDisplay.propTypes = {
  priceInWei: PropTypes.any,
  web3Service: PropTypes.any
};

export default ValueDisplay;
