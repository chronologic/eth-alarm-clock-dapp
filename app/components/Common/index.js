export { Step } from './Step';
export { InputField } from './InputField';
export { AwaitingMining } from './AwaitingMining';
export { Faucet } from './Faucet';
export { URLNotFound } from './URLNotFound';
export { ValueDisplay } from './ValueDisplay';
