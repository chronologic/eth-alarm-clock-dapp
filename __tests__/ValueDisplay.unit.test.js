import React from 'react';
import renderer from 'react-test-renderer';
import { ValueDisplay } from '../app/components/Common/ValueDisplay';
import { Provider } from 'mobx-react';
import Web3Service from '../app/services/web3';
import BigNumber from 'bignumber.js';

describe('ValueDisplay', () => {
  it('correctly displays prices', () => {
    const BIGNUMBERS_DISPLAY_MAP = [
      ['0', '0 ETH'],
      ['1', '1 WEI'],
      ['400', '400 WEI'],
      ['10000000', '0.00000000001 ETH'],
      ['40000000', '0.00000000004 ETH'],
      ['6000000000000000000000', '6000 ETH']
    ];

    const web3Service = new Web3Service();

    const injectables = {
      web3Service
    };

    for (let [bignumber, display] of BIGNUMBERS_DISPLAY_MAP) {
      let mockedRender = renderer.create(
        <Provider {...injectables}>
          <ValueDisplay priceInWei={new BigNumber(bignumber)} />
        </Provider>,
      );

      let tree = mockedRender.toJSON();
      expect(tree).toMatchSnapshot();

      expect(tree).toBe(display);
    }
  });
});