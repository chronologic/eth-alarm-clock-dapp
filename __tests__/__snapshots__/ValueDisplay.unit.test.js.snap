// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueDisplay correctly displays prices 1`] = `"0 ETH"`;

exports[`ValueDisplay correctly displays prices 2`] = `"1 WEI"`;

exports[`ValueDisplay correctly displays prices 3`] = `"400 WEI"`;

exports[`ValueDisplay correctly displays prices 4`] = `"0.00000000001 ETH"`;

exports[`ValueDisplay correctly displays prices 5`] = `"0.00000000004 ETH"`;

exports[`ValueDisplay correctly displays prices 6`] = `"6000 ETH"`;
